--- sys/netinet/ip_carp.c.orig  2025-11-28 03:00:00.000000000 +0300
+++ sys/netinet/ip_carp.c       2026-04-15 19:19:05.232426000 +0300
@@ -378,7 +378,22 @@

        return (0);
 }
+
+static u_short
+vrrp_cksum(const unsigned char *w, int len)
+{
+        u_short answer;
+        int sum = 0;

+        while (len-- > 0)
+                sum += *(w+len)<<8*(len % 2);
+
+        sum = (sum >> 16) + (sum & 0xffff);
+        sum += (sum >> 16);
+        answer = ~sum;
+        return (answer);
+}
+
 static void
 carp_hmac_prepare(struct carp_softc *sc)
 {
@@ -867,12 +882,12 @@
                goto out;
        }

-       if (carp_hmac_verify(sc, ch->carp_counter, ch->carp_md)) {
+       /*if (carp_hmac_verify(sc, ch->carp_counter, ch->carp_md)) {
                CARPSTATS_INC(carps_badauth);
                CARP_DEBUG("%s: incorrect hash for VHID %u@%s\n", __func__,
                    sc->sc_vhid, if_name(ifp));
                goto out;
-       }
+       }*/

        tmp_counter = ntohl(ch->carp_counter[0]);
        tmp_counter = tmp_counter<<32;
@@ -896,7 +911,7 @@
                 * If we receive an advertisement from a master who's going to
                 * be more frequent than us, go into BACKUP state.
                 */
-               if (timevalcmp(&sc_tv, &ch_tv, >) ||
+               if (timevalcmp(&sc_tv, &ch_tv, <) ||
                    timevalcmp(&sc_tv, &ch_tv, ==)) {
                        callout_stop(&sc->sc_ad_tmo);
                        carp_set_state(sc, BACKUP,
@@ -910,7 +925,7 @@
                 * If we're pre-empting masters who advertise slower than us,
                 * and this one claims to be slower, treat him as down.
                 */
-               if (V_carp_preempt && timevalcmp(&sc_tv, &ch_tv, <)) {
+               if (V_carp_preempt && timevalcmp(&sc_tv, &ch_tv, >)) {
                        carp_master_down_locked(sc,
                            "preempting a slower master");
                        break;
@@ -1075,10 +1090,8 @@
        } else
                sc->sc_counter++;

-       ch->carp_counter[0] = htonl((sc->sc_counter>>32)&0xffffffff);
-       ch->carp_counter[1] = htonl(sc->sc_counter&0xffffffff);
-
-       carp_hmac_generate(sc, ch->carp_counter, ch->carp_md);
+       bzero(ch->carp_counter,12);
+       ch->carp_counter[0] = (u_int32_t)(ifatoia(sc->sc_ifas[0])->ia_addr.sin_addr.s_addr);
 }

 static inline void
@@ -1223,7 +1236,7 @@
        ch.carp_vhid = sc->sc_vhid;
        ch.carp_advbase = sc->sc_advbase;
        ch.carp_advskew = advskew;
-       ch.carp_authlen = 7;    /* XXX DEFINE */
+       ch.carp_authlen = 1;
        ch.carp_pad1 = 0;       /* must be zero */
        ch.carp_cksum = 0;

@@ -1273,7 +1286,7 @@
                        goto resched;

                m->m_data += sizeof(*ip);
-               ch_ptr->carp_cksum = in_cksum(m, len - sizeof(*ip));
+               ch_ptr->carp_cksum = vrrp_cksum((unsigned char*)m->m_data, len - sizeof(*ip));
                m->m_data -= sizeof(*ip);

                CARPSTATS_INC(carps_opackets);
--- sys/netinet/ip_carp.h.orig  2025-11-28 03:00:00.000000000 +0300
+++ sys/netinet/ip_carp.h       2026-04-15 19:16:15.780718000 +0300
@@ -74,11 +74,10 @@
        u_int8_t        carp_pad1;      /* reserved */
        u_int8_t        carp_advbase;   /* advertisement interval */
        u_int16_t       carp_cksum;
-       u_int32_t       carp_counter[2];
-       unsigned char   carp_md[20];    /* SHA1 HMAC */
+       u_int32_t       carp_counter[3];
 } __packed;

-CTASSERT(sizeof(struct carp_header) == 36);
+CTASSERT(sizeof(struct carp_header) == 20);

 /*
  * The VRRPv3 header layout is as follows:
